tensorboard/backend/application.py

```diff
@@ -531,3 +562,58 @@ def _get_event_file_active_filter(flags):
   if inactive_secs < 0:
     return lambda timestamp: True
   return lambda timestamp: timestamp + inactive_secs >= time.time()
+
+
+class _DbModeMultiplexer(event_multiplexer.EventMultiplexer):
+  """Shim EventMultiplexer to use when in read-only DB mode.
+
+  In read-only DB mode, the EventMultiplexer is nonfunctional - there is no
+  logdir to reload, and the data is all exposed via SQL. This class represents
+  the do-nothing EventMultiplexer for that purpose, which serves only as a
+  conduit for DB-related parameters.
+
+  The load APIs raise exceptions if called, and the read APIs always
+  return empty results.
+  """
+  def __init__(self, db_uri, db_connection_provider):
+    """Constructor for `_DbModeMultiplexer`.
+
+    Args:
+      db_uri: A URI to the database file in use.
+      db_connection_provider: Provider function for creating a DB connection.
+    """
+    logger.info('_DbModeMultiplexer initializing for %s', db_uri)
+    super(_DbModeMultiplexer, self).__init__()
+    self.db_uri = db_uri
+    self.db_connection_provider = db_connection_provider
+    logger.info('_DbModeMultiplexer done initializing')
+
+  def AddRun(self, path, name=None):
+    """Unsupported."""
+    raise NotImplementedError()
+
+  def AddRunsFromDirectory(self, path, name=None):
+    """Unsupported."""
+    raise NotImplementedError()
+
+  def Reload(self):
+    """Unsupported."""
+    raise NotImplementedError()
+
+
+def _make_plugin_loader(plugin):
+  """Returns a plugin loader for the given plugin.
+
+  Args:
+    plugin: A TBPlugin subclass, or a TBLoader instance or subclass.
+
+  Returns:
+    A TBLoader for the given plugin.
+  """
+  if isinstance(plugin, base_plugin.TBLoader):
+    return plugin
+  if issubclass(plugin, base_plugin.TBLoader):
+    return plugin()
```

Comment:
This is a nice QOL change; I know that it’s bitten us in the past. Want
to update `tensorboard/default.py` to use this form for internal
consistency?

Reply:
Sure, but I'll also need to update `program.py` to support this form since that's actually where we do the expansion in the current code, and I'd rather not do that in this PR.

I can either leave this functionality in this path only, or I can back out the commit adding it and do the change to both places and `default.py` in a separate PR.  Any preference?

Down the road of course there should be just one path, but I was going to do more refactoring to pull out the contents of `standard_tensorboard_wsgi()` so that it can be invoked from `program.py` and then passed into `TensorBoardWSGIApp()` from there in a clean way.

Reply:
Ah, didn’t realize that this was an _additional_ path. Up to you.

Reply:
I've reverted this change for this PR and will send a followup for this separately.